### PR TITLE
Port regionWorker, step 1

### PR DESCRIPTION
**Pull request by xiaobogaga**
@coocood PTAL, thanks

### CODE REVIEW

File: tikv/raftstore/worker.go [hunk 1]
```diff
@@ -285,27 +289,182 @@ type pendingDeleteRanges struct {
...
+		}
+		task := r.pendingApplies[0].data.(*regionTask)
+		r.pendingApplies = r.pendingApplies[1:]
+		r.ctx.handleApply(task.regionId, task.status)
+	}
+}
+
+func (r *regionRunner) handleGen(regionTask *regionTask) {
+	go r.ctx.handleGen(regionTask.regionId, regionTask.raftSnapshot, regionTask.kvSnapshot, regionTask.notifier)
+	r.wg.Done()
```
Comment (coocood) [inline tikv/raftstore/worker.go]: `Done` should be called inside the goroutine.

File: tikv/raftstore/worker.go [hunk 2]
```diff
@@ -285,27 +289,182 @@ type pendingDeleteRanges struct {
...
+			break
+		}
+		task := r.pendingApplies[0].data.(*regionTask)
+		r.pendingApplies = r.pendingApplies[1:]
+		r.ctx.handleApply(task.regionId, task.status)
+	}
+}
+
+func (r *regionRunner) handleGen(regionTask *regionTask) {
+	go r.ctx.handleGen(regionTask.regionId, regionTask.raftSnapshot, regionTask.kvSnapshot, regionTask.notifier)
```
Comment (coocood) [inline tikv/raftstore/worker.go]: For now, we can just do it without goroutine.

File: tikv/raftstore/worker.go (hunk 1 above)
Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: yes, thanks.

File: tikv/raftstore/worker.go (hunk 2 above)
Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: ok

**Comment by xiaobogaga** [comment]
@coocood PTAL again, thanks

File: tikv/raftstore/worker.go [hunk 3]
```diff
@@ -285,14 +289,118 @@ type pendingDeleteRanges struct {
...
+// applySnap applies snapshot data of the Region.
+func (snapCtx *snapContext) applySnap(regionId uint64, status *JobStatus) (int, error) {
+	// todo, currently, it is a place holder.
+	return 0, nil
+}
+
+// handleApply tries to apply the snapshot of the specified Region. It calls `applySnap` to do the actual work.
+func (snapCtx *snapContext) handleApply(regionId uint64, status *JobStatus) {
+	atomic.CompareAndSwapInt64(status, JobStatus_Pending, JobStatus_Running)
+	retCode, err := snapCtx.applySnap(regionId, status)
```
Comment (coocood) [inline tikv/raftstore/worker.go]: We can only return the error, checks it against the predefined errors.

Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: ok

File: tikv/raftstore/worker.go [hunk 4]
```diff
@@ -50,11 +52,13 @@ type task struct {
...
 
 type regionTask struct {
-	regionId uint64
-	notifier chan<- *eraftpb.Snapshot
-	status   *JobStatus
-	startKey []byte
-	endKey   []byte
+	regionId     uint64
+	kvSnapshot   Snapshot
+	raftSnapshot Snapshot
```
Comment (coocood) [inline tikv/raftstore/worker.go]: We don't need `kvSnapshot` and `raftSnapshot`.

**Comment by xiaobogaga** [comment]
@coocood  PTAL again, thanks.

**Comment by xiaobogaga** [comment]
@coocood PTAL again, thanks.

**Comment by xiaobogaga** [comment]
@coocood PTAL again, thanks.

**Comment by coocood** [comment]
LGTM

**Pull request merged**